src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp

```diff
@@ -282,20 +281,22 @@ void QMCCostFunctionBatched::checkConfigurations(EngineHandle& handle)
   std::vector<int> samples_per_crowd_offsets(opt_num_crowds + 1);
   FairDivide(rank_local_num_samples_, opt_num_crowds, samples_per_crowd_offsets);
 
+
   // lambda to execute on each crowd
   auto evalOptConfig = [](int crowd_id, UPtrVector<CostFunctionCrowdData>& opt_crowds,
                           const std::vector<int>& samples_per_crowd_offsets, const std::vector<int>& walkers_per_crowd,
                           std::vector<ParticleGradient*>& gradPsi, std::vector<ParticleLaplacian*>& lapPsi,
                           Matrix<Return_rt>& RecordsOnNode, Matrix<Return_rt>& DerivRecords,
                           Matrix<Return_rt>& HDerivRecords, const SampleStack& samples, opt_variables_type& optVars,
                           bool needGrads, bool compute_nlpp, const std::string& includeNonlocalH,
-                          EngineHandle& handle) {
+                          EngineHandle& handle, int NumOptimizables) {
     CostFunctionCrowdData& opt_data = *opt_crowds[crowd_id];
 
     const int local_samples = samples_per_crowd_offsets[crowd_id + 1] - samples_per_crowd_offsets[crowd_id];
```

Comment:
It would be better to keep the arrays in LMYEngine sized as the full number of samples (per rank, in rank_local_num_samples_) and use samples_per_crowd_offsets to index into those arrays.   That should get rid of all (or most) of the thread calls in takeSample.

Reply:
takeSample has `ib` passed an argument, which is the crowd-local index.  It needs to be updated to also pass `is` ( supposed to be short for 'Index into Samples', not the word `is`), which is the index into samples on each rank.

Reply:
Both of those argument names could be long enough to carry information.  Two letters excepting those that are common words 'is', 'as', etc are fine for an iterated index in a small loop but not great as argument names.  `local_index` and `index_offset` come to mind.  Then only index offset would need non trivial docs.

Reply:
And I think sizing the arrays to the full number of samples outside of the concurrent lambda would do a good deal here to get a hacked first version, but every method called in the concurrent context and its calls etc needs to be evaluated for thread safety.  It's quite likely they will work fine in a single threaded context.